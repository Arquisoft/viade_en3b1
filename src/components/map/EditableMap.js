import React, { Component } from 'react';
import { Map, Polyline, TileLayer, Marker } from 'react-leaflet';

export class EditableMap extends Component {

    constructor() {
		super();
        this.state = { points: [] };
        this.firstPoint = true;
        
	}

    addPoint = (e) => {
		this.state.points.push(e.latlng);
        this.setState({ points: this.state.points.slice() });
        
        if(this.firstPoint === true) {
            this.props.openNotif("You can move a point by dragging it and delete it just clicking on it", 'info');
            this.firstPoint = false;
        }
    }

    updatePoint = (event) => {
		var id = event.target.options.marker_index;
		var position = event.target.getLatLng();
		const { points } = this.state;

		points[id] = position;
		this.setState({ points: points.slice() });
	}

	removePoint = (event) => {
        var id = event.target.options.marker_index;
        const { points } = this.state;
        points.splice(id, 1);
        this.setState({ points: points.slice() });

	}
    
    getTrackPoints() {
        return this.state.points.slice();
    }

    render() {
        const position = [43.3619145, -5.8493887];

        return (
            <React.Fragment>
                <Map
                    center={position}
                    zoom={12}
                    style={{ width: "52vh", height: "40vh" }}
                    onClick={this.addPoint}
                >
                    <TileLayer
                        url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
                        attribution='&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors'
                    />

                    <Polyline positions={this.state.points} color='orange' />

                    {this.state.points.map((position, index) =>
                        <Marker
                            marker_index={index}
                            position={position}
                            draggable={true}
                            ondrag={this.updatePoint}
                            onclick={this.removePoint}
                        >
                        </Marker>
                    )}
                </Map>
            </React.Fragment>
        );
    }
}

export default EditableMap;
