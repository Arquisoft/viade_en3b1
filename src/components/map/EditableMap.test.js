import EditableMap from "./EditableMap";
import { fireEvent, render, getByTestId } from "@testing-library/react";
import { Map } from "react-leaflet";

const mockOpenNotif = jest.fn();
mockOpenNotif.mockImplementation((text) => {
    return text;
});

test("Define EditableMap", () => {
    expect(EditableMap).toBeDefined();
});

test("Renders correct", () => {
    const wrapper = render(<EditableMap />);
    expect(wrapper).toMatchSnapshot();
});

test("Empty Map", () => {
    const wrapper = mount(<EditableMap />);
    expect(wrapper.instance().getTrackPoints().length).toEqual(0);
});

test("Click", () => {
    const wrapper = mount(<EditableMap />);
    const map = wrapper.find({id:"test-map-leaflet"}).first();
    map.simulate("click");
});

test("Add point", () => {
    const wrapper = mount(<EditableMap openNotif={mockOpenNotif} />);

    var latlng = L.latLng(50.5, 30.5);
    expect(wrapper.instance().getTrackPoints().length).toEqual(0);
    
    wrapper.instance().addPoint(latlng);

    expect(wrapper.instance().getTrackPoints().length).toEqual(1);
});

test("Remove point", () => {
    const wrapper = mount(<EditableMap openNotif={mockOpenNotif} />);

    var latlng = L.latLng(50.5, 30.5);
    expect(wrapper.instance().getTrackPoints().length).toEqual(0);
    
    wrapper.instance().addPoint(latlng);
    expect(wrapper.instance().getTrackPoints().length).toEqual(1);

    wrapper.instance().removePointById(0);
    expect(wrapper.instance().getTrackPoints().length).toEqual(0);
});

test("Update point", () => {
    const wrapper = mount(<EditableMap openNotif={mockOpenNotif} />);

    var latlng = L.latLng(1, 2);
    expect(wrapper.instance().getTrackPoints().length).toEqual(0);
    
    wrapper.instance().addPoint(latlng);
    expect(wrapper.instance().getTrackPoints().length).toEqual(1);

    var latlngNew = L.latLng(50, 50);
    wrapper.instance().updatePointById(0, latlngNew);
    expect(wrapper.instance().getTrackPoints().length).toEqual(1);
    expect(wrapper.instance().getTrackPoints()).toEqual([latlngNew]);
});

test("Remove point with click", () => {
    const wrapper = mount(<EditableMap openNotif={mockOpenNotif} />);

    var latlng = L.latLng(50.5, 30.5);
    expect(wrapper.instance().getTrackPoints().length).toEqual(0);
    
    wrapper.instance().addPoint(latlng);
    expect(wrapper.instance().getTrackPoints().length).toEqual(1);

    const map = wrapper.find({key:"0"}).first();
    map.simulate("click");

    expect(wrapper.instance().getTrackPoints().length).toEqual(0);
});