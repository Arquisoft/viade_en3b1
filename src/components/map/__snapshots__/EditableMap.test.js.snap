// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correct 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="leaflet-container leaflet-grab leaflet-touch-drag"
        id="test-map-leaflet"
        style="width: 34.5rem; height: 20rem; position: relative;"
        tabindex="0"
      >
        <div
          class="leaflet-pane leaflet-map-pane"
          style="left: 0px; top: 0px;"
        >
          <div
            class="leaflet-pane leaflet-tile-pane"
          >
            <div
              class="leaflet-layer "
              style="z-index: 1;"
            >
              <div
                class="leaflet-tile-container leaflet-zoom-animated"
                style="z-index: 18; left: 0px; top: 0px;"
              >
                <img
                  alt=""
                  class="leaflet-tile"
                  role="presentation"
                  src="https://a.tile.openstreetmap.org/12/1981/1499.png"
                  style="width: 256px; height: 256px; left: -114px; top: -109px;"
                />
              </div>
            </div>
          </div>
          <div
            class="leaflet-pane leaflet-shadow-pane"
          />
          <div
            class="leaflet-pane leaflet-overlay-pane"
          >
            <svg
              height="0"
              pointer-events="none"
              style="left: 0px; top: 0px;"
              viewBox="0 0 0 0"
              width="0"
            >
              <g>
                <path
                  class="leaflet-interactive"
                  d="M0 0"
                  fill="none"
                  stroke="#09012f"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-opacity="1"
                  stroke-width="3"
                />
              </g>
            </svg>
          </div>
          <div
            class="leaflet-pane leaflet-marker-pane"
          />
          <div
            class="leaflet-pane leaflet-tooltip-pane"
          />
          <div
            class="leaflet-pane leaflet-popup-pane"
          />
        </div>
        <div
          class="leaflet-control-container"
        >
          <div
            class="leaflet-top leaflet-left"
          >
            <div
              class="leaflet-control-zoom leaflet-bar leaflet-control"
            >
              <a
                aria-label="Zoom in"
                class="leaflet-control-zoom-in"
                href="#"
                role="button"
                title="Zoom in"
              >
                +
              </a>
              <a
                aria-label="Zoom out"
                class="leaflet-control-zoom-out"
                href="#"
                role="button"
                title="Zoom out"
              >
                −
              </a>
            </div>
          </div>
          <div
            class="leaflet-top leaflet-right"
          />
          <div
            class="leaflet-bottom leaflet-left"
          />
          <div
            class="leaflet-bottom leaflet-right"
          >
            <div
              class="leaflet-control-attribution leaflet-control"
            >
              <a
                href="https://leafletjs.com"
                title="A JS library for interactive maps"
              >
                Leaflet
              </a>
               | © 
              <a
                href="http://osm.org/copyright"
              >
                OpenStreetMap
              </a>
               contributors
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="leaflet-container leaflet-grab leaflet-touch-drag"
      id="test-map-leaflet"
      style="width: 34.5rem; height: 20rem; position: relative;"
      tabindex="0"
    >
      <div
        class="leaflet-pane leaflet-map-pane"
        style="left: 0px; top: 0px;"
      >
        <div
          class="leaflet-pane leaflet-tile-pane"
        >
          <div
            class="leaflet-layer "
            style="z-index: 1;"
          >
            <div
              class="leaflet-tile-container leaflet-zoom-animated"
              style="z-index: 18; left: 0px; top: 0px;"
            >
              <img
                alt=""
                class="leaflet-tile"
                role="presentation"
                src="https://a.tile.openstreetmap.org/12/1981/1499.png"
                style="width: 256px; height: 256px; left: -114px; top: -109px;"
              />
            </div>
          </div>
        </div>
        <div
          class="leaflet-pane leaflet-shadow-pane"
        />
        <div
          class="leaflet-pane leaflet-overlay-pane"
        >
          <svg
            height="0"
            pointer-events="none"
            style="left: 0px; top: 0px;"
            viewBox="0 0 0 0"
            width="0"
          >
            <g>
              <path
                class="leaflet-interactive"
                d="M0 0"
                fill="none"
                stroke="#09012f"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-opacity="1"
                stroke-width="3"
              />
            </g>
          </svg>
        </div>
        <div
          class="leaflet-pane leaflet-marker-pane"
        />
        <div
          class="leaflet-pane leaflet-tooltip-pane"
        />
        <div
          class="leaflet-pane leaflet-popup-pane"
        />
      </div>
      <div
        class="leaflet-control-container"
      >
        <div
          class="leaflet-top leaflet-left"
        >
          <div
            class="leaflet-control-zoom leaflet-bar leaflet-control"
          >
            <a
              aria-label="Zoom in"
              class="leaflet-control-zoom-in"
              href="#"
              role="button"
              title="Zoom in"
            >
              +
            </a>
            <a
              aria-label="Zoom out"
              class="leaflet-control-zoom-out"
              href="#"
              role="button"
              title="Zoom out"
            >
              −
            </a>
          </div>
        </div>
        <div
          class="leaflet-top leaflet-right"
        />
        <div
          class="leaflet-bottom leaflet-left"
        />
        <div
          class="leaflet-bottom leaflet-right"
        >
          <div
            class="leaflet-control-attribution leaflet-control"
          >
            <a
              href="https://leafletjs.com"
              title="A JS library for interactive maps"
            >
              Leaflet
            </a>
             | © 
            <a
              href="http://osm.org/copyright"
            >
              OpenStreetMap
            </a>
             contributors
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
